### PR TITLE
connectors-ci: auto retry on `DaggerError`

### LINKED ISSUE #28796: Transient build errors should be retried
Errors like `failed to authorize: failed to fetch oauth token: Post "https://auth.docker.io/token": EOF` and `failed to load cache key: Get "https://dtdg.co/latest-java-tracer": stream error: stream ID 3; INTERNAL_ERROR; received from peer` should not cause builds or connector tests to fail.  We should sleep for a few seconds and try again, up to 5 times.

Some examples:
* https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/nightly_builds/master/1690451195/9cee15b4db7988ceb7e9bbd4b85bcc26fface76e/destination-redshift/0.6.1/output.html
* https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/nightly_builds/master/1690451195/9cee15b4db7988ceb7e9bbd4b85bcc26fface76e/destination-bigquery/1.5.8/output.html

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes #28796

Our pipeline might occasionally face transient DaggerError (eg failure reaching out to DockerHub).
To avoid pipeline flakyness we can retry a step run on this type of failure

## How
* Make the retry logic more straightforward
* Add a `max_dagger_error_retries` attribute on `Step` which defaults to 3
* Test it

### CODE REVIEW

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`daec730fa3`](https://github.com/airbytehq/airbyte/pull/29081/commits/daec730fa3811b2843b0f704385a97948a36c271)) - ✅

⏲️  Total pipeline duration: 01mn31s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-openweather/metadata.yaml | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Code format checks                                                        | ✅        |
| Connector package install                                                 | ✅        |
| Build source-openweather docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                | ✅        |
| Acceptance tests                                                          | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_more-retries/1691151615/daec730fa3811b2843b0f704385a97948a36c271/source-openweather/0.1.6/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:daec730fa3811b2843b0f704385a97948a36c271)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by alafanechere** [comment]
Manual testing: 
* Ran the test pipeline on source-openweather successfully above ☝️ 
* [Successful pre-release of this connector](https://github.com/airbytehq/airbyte/actions/runs/5762148393/job/15621368449)

File: airbyte-ci/connectors/pipelines/pipelines/bases.py [hunk 1]
```diff
@@ -91,13 +91,16 @@ class Step(ABC):
 
     title: ClassVar[str]
     max_retries: ClassVar[int] = 0
+    max_dagger_error_retries: ClassVar[int] = 3
```
Comment (evantahler) [inline airbyte-ci/connectors/pipelines/pipelines/bases.py:94]: 3 retries - the classic starting place

File: airbyte-ci/connectors/pipelines/pipelines/bases.py [hunk 2]
```diff
@@ -91,13 +91,16 @@ class Step(ABC):
...
     should_log: ClassVar[bool] = True
     success_exit_code: ClassVar[int] = 0
     skipped_exit_code: ClassVar[int] = None
     # The max duration of a step run. If the step run for more than this duration it will be considered as timed out.
     # The default of 5 hours is arbitrary and can be changed if needed.
     max_duration: ClassVar[timedelta] = timedelta(hours=5)
 
+    retry_delay = timedelta(seconds=10)
```
Comment (evantahler) [inline airbyte-ci/connectors/pipelines/pipelines/bases.py:102]: I like how we can configure this as needed

**Review by evantahler — APPROVED** [review]

**Pull request merged**